src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace_ProjectReload.cs

```diff
@@ -0,0 +1,309 @@
+                }
+            }
+
+            if (lazyDocumentsToAdd != null)
+            {
+                project = addDocuments(project.Solution, lazyDocumentsToAdd.ToImmutable()).GetProject(project.Id)!;
+            }
+
+            // changed document
+            foreach (var (documentId, (documentInfo, newDocumentChecksums)) in newMap)
+            {
+                if (!oldMap.TryGetValue(documentId, out var oldDocumentChecksums))
+                {
+                    continue;
+                }
+
+                var document = project.GetDocument(documentId) ?? project.GetAdditionalDocument(documentId) ?? project.GetAnalyzerConfigDocument(documentId);
+                Contract.ThrowIfNull(document);
+
+                project = UpdateDocument(document, documentInfo, oldDocumentChecksums, newDocumentChecksums);
+            }
+
+            // removed document
+            foreach (var (documentId, _) in oldMap)
+            {
+                if (!newMap.ContainsKey(documentId))
+                {
+                    // we have a document removed
+                    project = removeDocument(project.Solution, documentId).GetProject(project.Id)!;
+                }
+            }
+
+            return project;
+        }
+
+        private static Project UpdateDocument(TextDocument document, DocumentInfo newDocumentInfo, DocumentStateChecksums oldDocumentChecksums, DocumentStateChecksums newDocumentChecksums)
+        {
+            // changed info
+            if (oldDocumentChecksums.Info != newDocumentChecksums.Info)
+            {
+                // there is no api to change these once document is created
+                Contract.ThrowIfFalse(document.State.Attributes.Id == newDocumentInfo.Id);
+                Contract.ThrowIfFalse(document.State.Attributes.Name == newDocumentInfo.Name);
+                Contract.ThrowIfFalse(document.State.Attributes.FilePath == newDocumentInfo.FilePath);
+                Contract.ThrowIfFalse(document.State.Attributes.IsGenerated == newDocumentInfo.IsGenerated);
+
+                if (document.State.Attributes.Folders != newDocumentInfo.Folders)
+                {
+                    // additional document can't change folder once created
+                    Contract.ThrowIfFalse(document is Document);
+                    document = document.Project.Solution.WithDocumentFolders(document.Id, newDocumentInfo.Folders).GetDocument(document.Id)!;
+                }
+
+                if (document.State.Attributes.SourceCodeKind != newDocumentInfo.SourceCodeKind)
+                {
+                    // additional document can't change sourcecode kind once created
+                    Contract.ThrowIfFalse(document is Document);
+                    document = document.Project.Solution.WithDocumentSourceCodeKind(document.Id, newDocumentInfo.SourceCodeKind).GetDocument(document.Id)!;
+                }
+            }
+
+            // we don't handle text changes here
```

Comment:
Is the idea that we'd have the file watching stuff done separately or something?